In [1]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split, cross_val_score, KFold, StratifiedKFold
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

In [2]:
wine = load_wine()
X, y = wine.data, wine.target

In [3]:
model = DecisionTreeClassifier(random_state=42)

## Part A

In [4]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

In [5]:
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

In [ ]:
acc_train_test_split = accuracy_score(y_test, y_pred)
print("Accuracy with Train-Test Split:", acc_train_test_split)

Accuracy with Train-Test Split: 0.9629629629629629
Variance with Train-Test Split: 0.6080246913580246


## Part B

In [14]:
kf = KFold(n_splits=5, shuffle=True, random_state=42)
scores_kf = cross_val_score(model, X, y, cv=kf)
print("K-Fold CV Accuracy (mean):", scores_kf.mean())
print('K-fold CV variance:', scores_kf.var())

K-Fold CV Accuracy (mean): 0.8763492063492062
K-fold CV variance: 0.0020525069286974037


In [15]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
scores_skf = cross_val_score(model, X, y, cv=skf)
print("Stratified K-Fold CV Accuracy (mean):", scores_skf.mean())
print("Stratified K-Fold CV Variance:", scores_skf.var())

Stratified K-Fold CV Accuracy (mean): 0.8931746031746032
Stratified K-Fold CV Variance: 0.0014461073318216156


In [16]:
from sklearn.model_selection import LeaveOneOut
loo = LeaveOneOut()
scores_loo = cross_val_score(model, X, y, cv=loo)
print("LOOCV Accuracy (mean):", scores_loo.mean())
print("LOOCV Variance:", scores_loo.var())

LOOCV Accuracy (mean): 0.8651685393258427
LOOCV Variance: 0.11665193788663047


## Analysis & Report

1. The Train-Test Split accuracy was **very high (0.96)**, but this may not reflect real-world performance since it depends on one random split.
2. Cross-validation (K-Fold, Stratified K-Fold,LOOCV) gave **lower but more reliable accuracies (~0.87–0.88)**. Which shows our model was more generalized.
3. The variance in CV scores of k-folds was **very less(~0.001-0.002)**, meaning the model performs consistently across different folds. Also it shows that k-fold trained models are more stable as compared to LOOCV trained
4. Conclusion: **Cross-validation is more reliable** for evaluating models compared to a single Train-Test split. Whereas **LOOCV is less stable** as compared to K-fold technique. 


## DataSet and Algorithm Used:
1. Wine Dataset (from sklearn.datasets)
2. Decision Tree Classifier (from sklearn.tree) (Because it tends to overfit on smaller datasets)